---
title: "PERMANOVA_FUT2"
author: "Yasmine"
date: "`r Sys.Date()`"
output: html_document
---

Purpose: determine associations between secretor status and microbiome composition. We are comparing non-secretors to secretor low fiber consumers (based on FFQ fiber here)

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(vegan)
library(microbiome)
library(phyloseq)
library(gtable)

```

```{r import}

# Genotyping 
fut2 <- read.csv("/Users/yasmine.bouzid/Documents/FUT2_analysis/data/forR_fut2.csv", header = T)

# Gut health / covar

df_l <- read.csv("/Users/yasmine.bouzid/Documents/GitHub/diet_guthealth/data/df_l.csv", header = T)

# Subset the dfs above with only what we'll need in the microbiome analysis (secretor status, age, sex, BMI)

fut2 <- fut2[ , -c(2:4)]
df_l <- df_l[ , c("subject_id", "kcal", "fibe", "age", "sex", "bmi_final")]

# MEEEERGE 

df <- merge(fut2, df_l, by = "subject_id") # n = 338

rm(df_l, fut2)

# unfortunately, need to correct structure of some variables each time -->

df$subject_id <- as.factor(as.character(df$subject_id))
df$sex <- df$sex %>% recode_factor('1' = 'Male', '2' = "Female")
df$secr_status <- as.factor(as.character(df$secr_status))

```

```{r wrangle-more}

# create fiber pk variable 
df$fibe_pk <- df$fibe/(df$kcal/1000)

# create df with only non-secretors 

df_ns <- df[df$secr_status == "non_secretor", ]
df_s <- df[df$secr_status == "secretor", ]

# Calculate the tertile cut points
fibe_pk_tertile_cut_points <- quantile(df_s$fibe_pk, probs = c(0, 1/3, 2/3, 1))

# Assign the tertile values to the dataframe
df_s$fibe_pk_tertile <- cut(df_s$fibe_pk, breaks = fibe_pk_tertile_cut_points, labels = c("T1", "T2", "T3"), include.lowest = TRUE)

# Check the result
head(df_s)


# adding to each to make NS and tertile values

missing_cols <- setdiff(names(df_s), names(df_ns))
print(missing_cols)

for (col in missing_cols) {
  df_ns[[col]] <- NA
}

# Ensure the columns are in the same order
df_ns <- df_ns[names(df_s)]

# Print the updated DataFrame 1
print(df_ns)


combined_df <- rbind(df_ns, df_s)

# Print the combined dataframe
print(combined_df)
names(combined_df)

# replace NAs in the tertile columns for non-secretors with NS

# Columns to replace NA values
cols_to_replace <- c("fibe_pk_tertile")
replacement_value <- "NS"  # The value to replace NA with

# Loop through the columns and replace NA values
for (col in cols_to_replace) {
  combined_df[[col]][is.na(combined_df[[col]])] <- replacement_value
  combined_df[[col]] <- as.factor(combined_df[[col]])  # Ensure the column is a factor
}

# Print the result
head(combined_df)

df <- combined_df
rm(combined_df, df_ns, df_s)

# we need to remove T2 and T3 from the dataframe 
# Remove rows where 'status' column equals 'closed' and 'category' column equals 'B'
df_filtered <- df[!df$fibe_pk_tertile %in% c("T2", "T3"), ]

# Print the result
print(df_filtered)

# drop factor levels 
df_filtered$fibe_pk_tertile <- droplevels(df_filtered$fibe_pk_tertile) # cheers


df <- df_filtered
rm(df_filtered)

```


Ok so here's where we try the PERMANOVA on metaphlan output...
https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/multivariate-comparisons-of-microbial-community-composition.html

```{r PERMANOVA-diff}

otu_table_data <- read.table("/Users/yasmine.bouzid/Documents/FUT2_analysis/data/merged_metaphlan_v4-0-6.txt", check.names = F, header = T, sep = "\t", row.names = 1)

# Create logical vectors
s_filter <- grepl("s__", row.names(otu_table_data))
t_filter <- !grepl("t__", row.names(otu_table_data))

# Combine the logical vectors to filter the data
otu_table_data <- otu_table_data[s_filter & t_filter, ]

# Convert the OTU table to a matrix and then to a phyloseq otu_table object
otu_table_matrix <- as.matrix(otu_table_data)
otu_table_phyloseq <- otu_table(otu_table_matrix, taxa_are_rows = TRUE)

# Read the metadata
row.names(df) <- df$subject_id
df <- df[ , -1]
metadata <- df

# Convert the metadata to a phyloseq sample_data object
sample_data_phyloseq <- sample_data(metadata)

# Check sample names in the OTU table
otu_sample_names <- colnames(otu_table_phyloseq)
print(otu_sample_names)

# Check sample names in the metadata
metadata_sample_names <- rownames(metadata)
print(metadata_sample_names)

# Ensure they match
all(otu_sample_names == metadata_sample_names) # they don't 

# Ensure sample names match
otu_sample_names <- colnames(otu_table_phyloseq)
metadata_sample_names <- rownames(metadata)

# Find common sample names
common_samples <- intersect(otu_sample_names, metadata_sample_names)

# Subset the OTU table and metadata to common samples
otu_table_phyloseq <- otu_table_phyloseq[, common_samples]
metadata <- metadata[common_samples, ]

# Check again to ensure sample names match
all(colnames(otu_table_phyloseq) == rownames(metadata))

# Convert the metadata to a phyloseq sample_data object
sample_data_phyloseq <- sample_data(metadata)

# Create the phyloseq object
pseq <- phyloseq(otu_table_phyloseq, sample_data_phyloseq)

# Check the phyloseq object
print(pseq)

# ...YAY
```

```{r transform-perm}

pseq.rel <- microbiome::transform(pseq, "compositional")
# hokay well apparrently this is supposed to be phyloseq object first ^
# I could cry, it worked. 

otu <- abundances(pseq.rel)
meta <- meta(pseq.rel)

# samples x species as input
permanova <- adonis2(t(otu) ~ fibe_pk_tertile,
               data = meta, permutations=99, method = "bray")

print(permanova) # no significant difference in microbiome by secretor status (P = 0.3)

```
```{r check-homogeneity}

# not sure if it's necessary since there's no difference but let's practice 

# Note the assumption of similar multivariate spread among the groups
# ie. analogous to variance homogeneity

dist <- vegdist(t(otu))
anova(betadisper(dist, meta$fibe_pk_tertile))

permutest(betadisper(dist, meta$fibe_pk_tertile), pairwise = TRUE)
# p = 0.59

```

```{r visualize}

# 

p <- plot_landscape(pseq.rel, method = "NMDS", distance = "bray", col = "secr_status", legend = T, size = 3) + 
  labs(title = "Non-metric Multi-dimensional Scaling Plot of Microbial Composition \n by Non-secretors and Secretor ASA24 Low Fiber Consumers",  x = "PC1",
       y = "PC2") +
#  scale_shape_discrete(labels = c("Non-secretor", "Secretor")) +
  scale_colour_brewer(labels = c("Non-secretor", "Secretor: Low Fiber"),
                      palette = "Dark2") +
  theme_bw() + 
  theme(
    text = element_text(size = 18),
    plot.title = element_text(size = 20),
    axis.title = element_text(size = 16),
    axis.text = element_text(size = 14),
    legend.position = "right",
    plot.margin = margin(10, 50, 10, 30, unit = "pt")
  )

print(p)

ggsave(filename = '/Users/yasmine.bouzid/Documents/FUT2_analysis/plots/permanova_tertileASA24low.jpg', plot = p, device = 'jpeg', dpi = 300, width = 10, height = 6, units = "in")

```

**Summary of Findings**: 

- No difference in microbiome between low fiber secretors and non-secretors
